logstash-core/src/test/java/org/logstash/common/io/ByteBufferPageIOTest.java

```diff
@@ -63,64 +64,217 @@ public void hasSpaceAfterWrite() throws IOException {
     }
 
     @Test
     public void recoversValidState() throws IOException {
         Queueable element = new StringElement("foobarbaz");
         long seqNum = 42L;
-        ByteBufferPageIO subject = subject();
-        subject.create();
-        subject.write(element.serialize(), seqNum);
+        ByteBufferPageIO io = newEmptyPageIO();
+        io.write(element.serialize(), seqNum);
 
-        byte[] inititalState = subject.dump();
-        subject = subject(inititalState.length, inititalState);
-        subject.open(seqNum, 1);
-        assertThat(subject.getElementCount(), is(equalTo(1)));
-        assertThat(subject.getMinSeqNum(), is(equalTo(seqNum)));
+        byte[] inititalState = io.dump();
+        io = newPageIO(inititalState.length, inititalState);
+        io.recover();
+        assertThat(io.getElementCount(), is(equalTo(1)));
+        assertThat(io.getMinSeqNum(), is(equalTo(seqNum)));
     }
 
     @Test(expected = IOException.class)
-    public void recoversInvalidState() throws IOException {
+    public void openUnexpectedSeqNum() throws IOException {
         Queueable element = new StringElement("foobarbaz");
         long seqNum = 42L;
-        ByteBufferPageIO subject = subject();
-        subject.create();
-        subject.write(element.serialize(), seqNum);
+        ByteBufferPageIO io = newEmptyPageIO();
+        io.write(element.serialize(), seqNum);
 
-        byte[] inititalState = subject.dump();
-        subject(inititalState.length, inititalState);
-        subject.open(1L, 1);
+        byte[] inititalState = io.dump();
+        newPageIO(inititalState.length, inititalState);
+        io.open(1L, 1);
     }
 
-    // TODO: add other invalid initial states
+    private interface BufferGenerator {
+        byte[] generate() throws IOException;
+    }
+
+    private final List<BufferGenerator> singleGenerators = Arrays.asList(
+        // invalid length
+        () -> {
+            Queueable element = new StringElement("foobarbaz");
+            ByteBufferPageIO io = newEmptyPageIO();
+            byte[] bytes = element.serialize();
+            io.write(bytes, 1L, 514, io.checksum(bytes));
+            return io.dump();
+        },
+
+        // invalid checksum
+        () -> {
+            Queueable element = new StringElement("foobarbaz");
+            ByteBufferPageIO io = newEmptyPageIO();
+            byte[] bytes = element.serialize();
+            io.write(bytes, 1L, bytes.length, 77);
+            return io.dump();
+        },
+
+        // invalid payload
+        () -> {
+            Queueable element = new StringElement("foobarbaz");
+            ByteBufferPageIO io = newEmptyPageIO();
+            byte[] bytes = element.serialize();
+            int checksum = io.checksum(bytes);
+            bytes[1] = 0x01;
+            io.write(bytes, 1L, bytes.length, checksum);
+            return io.dump();
+        }
+    );
+
+    private final List<BufferGenerator> doubleGenerators = Arrays.asList(
```

Comment:
This would probably be better written using JUnit parameterized tests for the generated dumps.

```java
@RunWith(Parameterized.class)
public class MyTestClass {
    @Parameters
    public static Collection<byte[]> data() {
       ... generate all `io.dump()` calls here and return a Collection<> of them.
    }

    @Parameter
    public byte[] bytes;

    @Test
    public void openInvalidElement() {
         ByteBufferPageIO io = newPageIO(bytes.length, bytes);
         io.open(1L, 1);

         assertThat(io.getElementCount(), is(equalTo(1)));
         assertThat(io.getMinSeqNum(), is(equalTo(1L))); 
    }
}
```

This has nice properties of avoiding the iteration you do on singleGenerators and doubleGenerators. You could also use this to simplify the invalid write generation.